src/mcp/http/client-registration.ts

```diff
@@ -0,0 +1,191 @@
+      error_description: this.message,
+    }
+  }
+}
+
+/**
+ * Check if a request body is an OAuth Dynamic Client Registration request.
+ *
+ * @param body - The request body to check
+ * @returns True if the body looks like a client registration request
+ */
+export function isClientRegistrationRequest(body: unknown): body is ClientRegistrationRequest {
+  if (!body || typeof body !== 'object') return false
+
+  const obj = body as Record<string, unknown>
+  return !!(
+    obj.redirect_uris ||
+    obj.grant_types ||
+    obj.token_endpoint_auth_method ||
+    obj.response_types ||
+    obj.client_name
+  )
+}
+
+/**
+ * Validate a client registration request.
+ *
+ * @param request - The client registration request
+ * @throws ClientRegistrationError if validation fails
+ */
+export function validateClientRegistrationRequest(request: ClientRegistrationRequest): void {
+  // redirect_uris is required and must be a non-empty array
+  if (
+    !request.redirect_uris ||
+    !Array.isArray(request.redirect_uris) ||
+    request.redirect_uris.length === 0
+  ) {
+    throw new ClientRegistrationError(
+      'invalid_request',
+      'redirect_uris is required and must be a non-empty array',
+    )
+  }
+
+  // Validate each redirect_uri is a valid URL
+  for (const uri of request.redirect_uris) {
+    try {
+      new URL(uri)
+    } catch {
+      throw new ClientRegistrationError('invalid_redirect_uri', `Invalid redirect_uri: ${uri}`)
+    }
+  }
+
+  // Validate grant_types if provided
+  const validGrantTypes = ['authorization_code', 'refresh_token', 'client_credentials']
+  if (request.grant_types) {
+    for (const grantType of request.grant_types) {
+      if (!validGrantTypes.includes(grantType)) {
+        throw new ClientRegistrationError(
+          'invalid_client_metadata',
+          `Unsupported grant_type: ${grantType}`,
+        )
+      }
+    }
+  }
+
+  // Validate response_types if provided
+  const validResponseTypes = ['code', 'token']
+  if (request.response_types) {
+    for (const responseType of request.response_types) {
+      if (!validResponseTypes.includes(responseType)) {
+        throw new ClientRegistrationError(
+          'invalid_client_metadata',
+          `Unsupported response_type: ${responseType}`,
+        )
+      }
+    }
+  }
+
+  // Validate token_endpoint_auth_method if provided
+  const validAuthMethods = ['none', 'client_secret_basic', 'client_secret_post']
+  if (
+    request.token_endpoint_auth_method &&
+    !validAuthMethods.includes(request.token_endpoint_auth_method)
+  ) {
+    throw new ClientRegistrationError(
+      'invalid_client_metadata',
+      `Unsupported token_endpoint_auth_method: ${request.token_endpoint_auth_method}`,
+    )
+  }
+}
+
+/**
+ * Generate a cryptographically secure client secret.
+ *
+ * @returns A base64url-encoded random string
+ */
+async function generateClientSecret(): Promise<string> {
+  const crypto = await import('crypto')
```

Comment:
is there a reason to use an await import here instead of just importing the crypto library at the top of the file?